### ISSUE TITLE
Speculative decoding does not reproduce default generation when presence/frequency penalties are active

### ISSUE THREAD

**Issue by rcfa** [osaurus-ai/vmlx-swift#357]
`SpeculativeDecodingTests."Speculative decoding matches default token generation"` fails on a
clean checkout for all three `withLogitProcessor: true` cases (draft sizes 2, 8, 48). All three
`false` cases pass, so the speculative path itself is sound and the divergence is specific to the
logit processors.

## Which penalty

Isolated per penalty, at `numDraftTokens: 8`:

| penalty | matches default |
|---|---|
| `repetitionPenalty` | yes |
| `presencePenalty` | **no** |
| `frequencyPenalty` | **no** |

That split is predicted by the storage. `RepetitionContext` holds a `TokenRing`, whose `append`
REASSIGNS its `MLXArray` buffer, so an ordinary struct copy is genuinely independent.
`PresencePenaltyContext` and `FrequencyPenaltyContext` hold `GeneratedTokenCounts`, which is a
`final class` mutated in place by `record`.

## One cause found and fixed

`SpeculativeTokenIterator` copies the processor twice per round — `var draftProcessor = processor`
("Copy to discard later") and `if var verifyProcessor = processor`. With a reference-typed member
neither copy discards anything: the draft loop records every PROPOSED token and the verify loop walks
all `numDraft + 1` positions, so rejected drafts are recorded into the real processor and accepted
ones are recorded twice.

Fixed by adding `LogitProcessor.independentCopy()` (default = the value copy, overridden by the two
penalty contexts and `PenaltyProcessor` to deep-copy the counts object and its array), and using it
at both copy sites. `LogitProcessorIndependentCopyTests` covers the mechanism directly: recording
into a copy no longer reaches the original, and the copy still accumulates its own penalty.

## A second mechanism remains

With that fixed and the copy provably independent, presence and frequency STILL diverge. So there is
a second cause I have not identified. Filing rather than guessing.

Both tests are wrapped in `withKnownIssue` for the penalty cases only, so the suite is green and the
day this starts passing it reports as an unexpected pass rather than going unnoticed — the divergence
was previously masked because the whole suite was permanently red.

**Comment by rcfa** [comment]
Found the second mechanism. Both halves are now fixed and the suite passes with no known issues.

**`GeneratedTokenCounts.record` was silently dropping tokens.** It is handed a token, not the logits, so it does not know the vocabulary size and cannot allocate — and it simply returned:

```swift
guard let counts else { return }   // no logits seen yet, so nothing to count into
```

On the ordinary decode path that is invisible, because `process` runs before every `didSample` and allocates on the first step. On the speculative path the real processor is advanced with `didSample` while only its throwaway copies ever see logits — so it can be handed many tokens before it sees any. Instrumented: **64 of 1106 `record` calls were dropped** in this test.

Fix: hold those tokens instead of dropping them, and fold them in the moment `vector` can size the buffer. `independentCopy` carries the pending list too.

With that and the earlier deep-copy fix, all six `Speculative decoding matches default token generation` cases pass, and the per-penalty test passes for repetition, presence and frequency. The `withKnownIssue` wrappers are removed — and they earned their keep: the first run after this fix reported "Known issue was NOT recorded", which is how I knew it was complete rather than partly better.

Closing.
